### PR TITLE
James 2361 parse mdn report part 1

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+
+/**
+ * <p>This matcher checks if the content type matches.</p>
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java]: Isn't it ContentType parameters?

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
+package org.apache.james.transport.matchers;
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: Missing license :trollface:

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                        .collect(Guavate.toImmutableList());
+                } catch (IllegalArgumentException e) {
+                    throw new MailetException("error parsing configuration", e);
+                }
+            }
+        ).sneakyThrow();
+    }
+
+    private Pair<String, String> conditionToPair(String s) {
+        Preconditions.checkArgument(s.contains("="));
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java]: Is `abc=` valid?

**Comment by chibenwa** [comment]
test this please

**Comment by rouazana** [comment]
```
[c0e3bb24174ebd875ea18d094494abd86a51b72d] 	Can't use this cluster instance because it was previously closed
[c0e3bb24174ebd875ea18d094494abd86a51b72d] 	<no message> in java.lang.NullPointerException
```
test this please

**Comment by chibenwa** [comment]
```
[c0e3bb24174ebd875ea18d094494abd86a51b72d] Tests in error: 
[c0e3bb24174ebd875ea18d094494abd86a51b72d]   ICSAttachmentWorkflowTest.yahooBase64CalendarAttachmentShouldBePublishedInMQWhenMatchingWorkflowConfiguration:707 » ConditionTimeout
```

test this please

**Comment by rouazana** [comment]
pushed force to rebase on master

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java (hunk 2 above)
Comment (mbaechler) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: if you need a proof that intellij doesn't like me

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java [hunk 4]
```diff
@@ -45,10 +47,9 @@ void setUp() throws MessagingException {
         sampleMail = FakeMail.builder()
             .mimeMessage(MimeMessageBuilder
                 .mimeMessageBuilder()
-                .addHeader("Content-Type",
-                    "multipart/report; report-type=\"disposition-notification\"; boundary=\"=-ac61K8KXSRpaQ/eveStc\"")
                 .setSubject("Mail read")
-                .setText("You email has been read by Bart"))
+                .addHeader("Content-Type", TEST_CONTENT_TYPE)
+                .setText("You email has been read by Bart", TEST_CONTENT_TYPE))
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: why do you need to put the content type here ?

Comment (rouazana) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: else it creates the message with text/plain mimetype

Comment (mbaechler) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: maybe we should not call addHeader before, then ?

Comment (rouazana) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/HasMimeTypeParameterTest.java]: I also tried, but it fails without it. Feel free to debug it :)

File: mdn/src/main/java/org/apache/james/mdn/MDN.java [hunk 5]
```diff
@@ -93,8 +94,9 @@ public MDNReport getReport() {
     }
 
     public MimeMultipart asMultipart() throws MessagingException {
-        MimeMultipart multipart = new MimeMultipart();
+        MimeMultipartReport multipart = new MimeMultipartReport();
         multipart.setSubType(REPORT_SUB_TYPE);
+        multipart.setReportType("disposition-notification");
```
Comment (aduprat) [inline mdn/src/main/java/org/apache/james/mdn/MDN.java]: We may introduce an enum for report type.

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+ *     <code>
+ *         <mailet match="HasMimeTypeParameter=report-type=disposition-notification,report-type=other" class="..." />
+ *     </code>
+ * </pre>
+ */
+public class HasMimeTypeParameter extends GenericMatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HasMimeTypeParameter.class);
+
+    @VisibleForTesting List<Pair<String, String>> filteredMimeTypeParameters;
```
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java]: `final` missing

File: mdn/src/main/java/org/apache/james/mdn/MDN.java (hunk 5 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDN.java]: Hmmm I more prefer a static field for this. (An enum makes sens IMO in the Mime handling library? Here it is javax.mail so we cannot change it. and do not allow much extension. Do we know all the possible report types?

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java (hunk 6 above)
Comment (chibenwa) [reply mailet/standard/src/main/java/org/apache/james/transport/matchers/HasMimeTypeParameter.java]: For a field that can only be setted up in `init()` and not in the constructor?

**Comment by chibenwa** [comment]
```
[9260edb51478c899ea25accce481432ad759c5d5] Tests in error: 
[9260edb51478c899ea25accce481432ad759c5d5] Feature: An upload endpoint should be available to upload contents.Scenario: Uploading a content without being authenticated
[9260edb51478c899ea25accce481432ad759c5d5]   Run 1: MemoryUploadCucumberTest » Socket Broken pipe (Write failed)
[9260edb51478c899ea25accce481432ad759c5d5]   Run 2: MemoryUploadCucumberTest » Socket Broken pipe (Write failed)
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**